### ISSUE TITLE
proposal: check for import path collision

### ISSUE THREAD

**Issue by UsernameIsAlreadyTaken6** [golang/go#26227]
### What version of Go are you using (`go version`)?
go version go1.10.3 linux/amd64

### Does this issue reproduce with the latest release?
Yes

**Proposal:** Raise a warning or an error at compile time if the import path of a local package matches the import path of package from the standard library.
This is prevents a fairly niche problem, namely collisions, but in practice the compile error that they produce remind of a declaration/syntax/code organization error so it's pretty hard to guess what went wrong.

**Comment by UsernameIsAlreadyTaken6** [comment]
When the collision happened I've got compile-time errors such as:
> undefined: pkgname.XXXXXX".

Maybe a good compromise would be to check if "pkgname" matches the name of package library when these errors happen, and if that the case,  print a final error "No XXXXXX provided by standard library package pkgname"?

**Comment by mvdan** [comment]
Perhaps so. I'm not sure how smart should the compiler errors get. /cc @mdempsky @randall77 @josharian

**Comment by josharian** [comment]
If we did this it would be in cmd/go instead of cmd/compile. But this is a legitimate thing to do, so I don’t think we should make it an error. Also, I have personally used vendored versions of modified stdlib packages to incorporate fixes and performance improvements that were not upstream yet.

**Comment by mdempsky** [comment]
As best I can tell, the issue is that @UsernameIsAlreadyTaken6 created a $GOPATH/src/foo/bar package; but then when they tried to use it, $GOROOT/src/foo/bar was used instead?

In practice, this doesn't happen because user import paths are rooted in a domain name, which contains a period, whereas stdlib packages don't contain periods. Maybe we should formalize that practice better with a vet or lint warning?

**Comment by UsernameIsAlreadyTaken6** [comment]
I thought naming my package "context" would be nice (I'm a newbie).. turns out it was not very nice.
I didn't use a domain name because it's a local project at the moment and deliberately adding a dummy domain would have been weird, but.. that's your call really.

**Comment by rsc** [comment]
If you tried to run "go install" in your context dir you get a good error:

```
$ cd $GOPATH/src
$ mkdir context
$ cd context
$ echo package context >c.go
$ go install
go install: no install location for /Users/rsc/src/context: hidden by /Users/rsc/go/src/context
$ 
```

(/Users/rsc/go is my $GOROOT.)

Maybe we should break "go build" too? I don't think we should make all builds trying to import the standard context break just because $GOPATH/src/context exists. And soon GOPATH will be gone anyway.

**Comment by bradfitz** [comment]
Seems like making `go build` and `go install` consistent in their error messages makes sense.

**Comment by mknyszek** [comment]
This came up again as the lack of an error leaks into downstream tooling and makes for a subpar user experience. See #53825. That was closed as a duplicate, but we may want to consider raising the priority of resolving this. Unclear who should do it, though.

**Comment by mvdan** [comment]
> does this mean each new package in the std library is breaking backwards compatibility?

No, given https://github.com/golang/go/issues/32819. As you correctly point out, without reserving dot-less import paths for the standard library, adding any new standard library package could potentially break users.
